src/pgen/torus.cpp

```diff
@@ -534,15 +547,115 @@ void PostInitializationModifier(ParameterInput *pin, Mesh *pmesh) {
   for (auto &pmb : pmesh->block_list) {
     auto &rc = pmb->meshblock_data.Get();
     auto geom = Geometry::GetCoordinateSystem(rc.get());
+    auto coords = pmb->coords;
+
+    // set up transformation stuff
+    auto gpkg = pmb->packages.Get("geometry");
+    bool derefine_poles = gpkg->Param<bool>("derefine_poles");
+    Real h = gpkg->Param<Real>("h");
+    Real xt = gpkg->Param<Real>("xt");
+    Real alpha = gpkg->Param<Real>("alpha");
+    Real x0 = gpkg->Param<Real>("x0");
+    Real smooth = gpkg->Param<Real>("smooth");
+    auto tr = Geometry::McKinneyGammieRyan(derefine_poles, h, xt, alpha, x0, smooth);
 
     auto ib = pmb->cellbounds.GetBoundsI(IndexDomain::interior);
     auto jb = pmb->cellbounds.GetBoundsJ(IndexDomain::interior);
     auto kb = pmb->cellbounds.GetBoundsK(IndexDomain::interior);
 
     PackIndexMap imap;
-    auto v = rc->PackVariables({fluid_prim::bfield}, imap);
+    std::vector<std::string> vars = {fluid_prim::bfield, fluid_prim::density};
+    auto v = rc->PackVariables(vars, imap);
     const int iblo = imap[fluid_prim::bfield].first;
     const int ibhi = imap[fluid_prim::bfield].second;
+    const int irho = imap[fluid_prim::density].first;
+
+    auto tracer_pkg = pmb->packages.Get("tracers");
+    bool do_tracers = tracer_pkg->Param<bool>("active");
+
+    /* tracer initialization. lots of logic to ensure we only put tracers on disk */
+    if (do_tracers) {
+      auto &sc = pmb->swarm_data.Get();
+      auto &swarm = pmb->swarm_data.Get()->Get("tracers");
+      auto rng_pool = tracer_pkg->Param<RNGPool>("rng_pool");
+      const auto num_tracers_total = tracer_pkg->Param<int>("num_tracers");
+      auto &x = swarm->Get<Real>("x").Get();
+      auto &y = swarm->Get<Real>("y").Get();
+      auto &z = swarm->Get<Real>("z").Get();
+      auto &mass = swarm->Get<Real>("mass").Get();
+      auto &id = swarm->Get<int>("id").Get();
+
+      /* lambda to flatten (k, j, i) -> n */
+      auto flatten = KOKKOS_LAMBDA(const int k, const int j, const int i) {
+        return (i + (ib.e - ib.s + 1) * (j + (jb.e - jb.s + 1) * k));
+      };
+
+      // TODO (BLB): better way to construct & access indices array?
+      // Access as indices( flatten( k-kb.s, j-jb.s, i-ib.s ) )
```

Comment:
I like @brryan 's idea of using a "volume-weighted" mask, as discussed on MM. We should look at the tracers example parthenon.

Reply:
Note we don't have to fix this now. PR can go in even if the initialization is rough around the edges.